tracer/src/Datadog.Trace/Baggage.cs

```diff
@@ -0,0 +1,398 @@
+        {
+            lock (list)
+            {
+                for (int i = 0; i < list.Count; i++)
+                {
+                    // match both key and value
+                    if (list[i].Key == item.Key && list[i].Value == item.Value)
+                    {
+                        list.RemoveAt(i);
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    void ICollection<KeyValuePair<string, string>>.CopyTo(KeyValuePair<string, string>[] array, int arrayIndex)
+    {
+        if (array == null!)
+        {
+            ThrowHelper.ThrowArgumentNullException(nameof(array));
+        }
+
+        if (arrayIndex < 0 || arrayIndex >= array.Length)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(arrayIndex));
+        }
+
+        // check if items fit into the array
+        if (array.Length - arrayIndex < Count)
+        {
+            ThrowHelper.ThrowArgumentException(
+                """
+                The number of elements in the source collection is greater than 
+                the available space from arrayIndex to the end of the destination array.
+                """);
+        }
+
+        if (_items is { Count: > 0 } list)
+        {
+            lock (list)
+            {
+                for (int i = 0; i < list.Count; i++)
+                {
+                    var item = list[i];
+                    array[arrayIndex + i] = item;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the baggage value associated with the specified name.
+    /// </summary>
+    /// <param name="name">The baggage item name.</param>
+    /// <returns>Returns the baggage item value, or <c>null</c> if not found.</returns>
+    public string? GetValueOrDefault(string name)
+    {
+        if (name == null!)
+        {
+            ThrowHelper.ThrowArgumentNullException(nameof(name));
+        }
+
+        return TryGetValue(name, out var value) ? value : null;
+    }
+
+    /// <summary>
+    /// Removes all baggage items.
+    /// </summary>
+    public void Clear()
+    {
+        if (_items is { } list)
+        {
+            lock (list)
+            {
+                _items?.Clear();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Adds the baggage items from this baggage instance into <paramref name="destination"/>.
+    /// </summary>
+    public void MergeInto(Baggage destination)
+    {
+        if (destination == null!)
+        {
+            ThrowHelper.ThrowArgumentNullException(nameof(destination));
+        }
+
+        if (Count == 0)
+        {
+            // nothing to merge
+            return;
+        }
+
+        var sourceItems = _items!; // if count > 0, then _items is not null
+        var destinationItems = destination.EnsureListInitialized();
+
+        lock (sourceItems)
+        {
+            lock (destinationItems)
+            {
```

Comment:
there's an edge case that can cause a deadlock in this double lock.
eg:
```
Thread 1 => baggageA.MergeInto(baggageB);
Thread 2 => baggageB.MergeInto(baggageA);
```
both will adquire their first lock, over their `items` but then the deadlock occurs trying to acquire the items from the other baggage.

I realised this while working on a race detection algorithm, "one solution" for this case is to ensure the same lock ordering no matter the instance, you can base that on the hascode perhaps?
something like:
```
var lockObj1 = sourceItems.GetHashCode() < destinationItems.GetHashCode() ? sourceItems : destinationItems;
var lockObj2 = lockObj1 == sourceItems ? destinationItems : sourceItems;

lock (lockObj1)
lock (lockObj2)
```
That way is guaranteed that both calls will have the same lock order, so no deadlock will happen.

If you don't want to use the hash code you can create a private baggageId on the .ctor, so you can use later for the comparison.

Reply:
Double locks are evil 😆

Reply:
Does this mean that the lock in `AddOrReplace` should be removed? The `lock` calls everywhere has me 😵

Reply:
> The `lock` calls everywhere has me 😵

Seriously, haha. I can just remove them all and let users handle it just as if they were using `List<T>` or `Dictionary<TKey, TValue>` directly. Thread-safety is not generally assumed in .NET unless explicitly documented.

Reply:
Anyway, I fixed this one. Thanks!

Reply:
> Does this mean that the lock in `AddOrReplace` should be removed? 

The instance methods `AddOrReplace(string key, string value)` and `MergeInto(Baggage destination)` both call the static `AddOrReplace(List<KeyValuePair<string, string>> items, string key, string value)`. There is no locking inside the static `AddOrReplace()` since both callers take care of that.

I'll rename that static one so avoid confusion with the instance version.

Reply:
Fixed the possible deadlock and renamed the static `AddOrReplace`. Thanks!